torch/_functorch/aot_autograd.py

```diff
@@ -742,7 +744,10 @@ def from_fun(t):
     @wraps(f)
     def inner(*flat_args):
         # This function is meant to be run with the forward, which expects a flat list of tensor/symint/other args.
-        assert all(isinstance(a, KNOWN_TYPES) for a in flat_args)
+        def valid_input(a):
+            # Insanely sus, talk to folks about this
+            return isinstance(a, KNOWN_TYPES) or callable(flat_args[-1])
```

Comment:
Yeah I guess I still really don't understand why we want hook functions to be explicit arguments to the FX graph. I would almost prefer it if there was some extralingual mechanism for getting at them. @jansel could you walk me through the reasoning again?

Reply:
Yeah, we could come up with some other way I guess, but this makes it an easy stepping stone towards @jansel's plan of not specializing on function ids in favor of just specializing on position + count.

Reply:
Is there a discussion somewhere I can read on that?

It's not clear to me why making hook fns inputs into the FX graph will help you avoid specializing on fn ID's. If a user changes the hook that they register onto an intermediate, won't we still need to recompile, which would require guarding on the id of the actual hook fn?

e.g.:
```
@torch.compile
def fn_with_custom_hook(inp, hook):
    intermed = inp * 2
    intermed.register_hook(hook)
    out = intermed * 2
    return out
    
inp = torch.randn(4, requires_grad=True)
out1 = fn_with_custom_hook(inp, hook1)
# This should probably recompile?
out2 = fn_with_custom_hook(inp, hook2)
```

Reply:
Discussed offline